strawberryfields/tdm/tdmprogram.py

```diff
@@ -81,69 +85,176 @@ def validate_measurements(circuit, N):
     # check if all lists are of equal length
     param_lengths = [len(param) for param in args]
     if len(set(param_lengths)) != 1:
         raise ValueError("Gate-parameter lists must be of equal length.")
 
-    # check copies
-    if not isinstance(copies, int) or copies < 1:
-        raise TypeError("Number of copies must be a positive integer.")
 
+def _get_mode_order(num_of_values, modes, N, timebins):
+    """Get the order by which the modes were measured
+
+    The mode order is determined by the circuit and the mode-shifting occurring in
+    :class:`~.TDMProgram`. For the following circuit, the mode order returned by this
+    function would be [0, 2, 0, 1], duplicated shots number of times:
+
+    >>> prog = sf.TDMProgram(N = [1, 2])
 
-def _get_mode_order(num_of_values, N):
-    """Get the order by which the modes were measured"""
+    >>> with prog.context([1, 2], [4, 5]) as (p, q):
+    ...     MeasureHomodyne(p[0]) | q[0]
+    ...     MeasureHomodyne(p[1]) | q[2]
+
+    """
     all_modes = []
     for i in range(len(N)):
-        ra = list(range(sum(N[:i]), sum(N[: i + 1])))
-        all_modes.append(ra * ceil(max(N) / len(ra)))
-
+        timebin_modes = list(range(sum(N[:i]), sum(N[: i + 1])))
+        # shift the timebin_modes if the measured mode isn't the first in the
+        # band, so that the measurements start at the correct mode
+        shift = modes[i] - sum(N[:i])
+        timebin_modes = timebin_modes[shift:] + timebin_modes[:shift]
+
+        # extend the modes by duplicating the list so that the measured mode
+        # orders in all bands have the same length
+        extended_modes = timebin_modes * ceil(1 + timebins // len(timebin_modes))
+        all_modes.append(extended_modes[:timebins])
+
+    # alternate measurements in the bands and extend/duplicate the resulting
+    # list so that it is at least as long as `num_of_values`
     mode_order = [i for j in zip(*all_modes) for i in j]
-    mode_order *= ceil(num_of_values / len(mode_order))
+    mode_order *= ceil(1 + num_of_values / len(mode_order))
 
     return mode_order[:num_of_values]
 
 
-def reshape_samples(all_samples, modes, N):
+def reshape_samples(all_samples, modes, N, timebins):
     """Reshapes the samples dict so that they have the expected correct shape.
 
-    Corrects the :attr:`~.Results.all_samples` dictionary so that the measured modes are the ones
-    defined to be measured in the circuit, instead of being spread over a larger
+    Corrects the :attr:`~.Results.all_samples` dictionary so that the measured modes are
+    the ones defined to be measured in the circuit, instead of being spread over a larger
     number of modes due to the mode-shifting occurring in :class:`~.TDMProgram`.
 
+    The function iterates through samples obtained from the unrolled circuit to populate
+    and return a new samples dictionary with the shape ``{spatial mode: (shots,
+    timebins)}``. E.g., this unrolled circuit:
+
+    .. code-block:: pycon
```

Comment:
This codeblock (and the two below) can probably be changed to `python`, since there is no python console syntax present